Request code review for rspec  branch

javitocor opened the pull request:
In this branch, using Rspec testing framework we have added spec to check the status of the game in all conditions (win, tie or still playing).

SpaYco committed:
added RSpec tests

SpaYco committed:
added game_start.rb to start the game with

javitocor committed:
Fixed rubocop errors

SpaYco commented:
Please Note: there's no issues with rubocop on our end, but for some reason stickler says otherwise
 
![image](https://user-images.githubusercontent.com/11758151/77074371-0f1fd280-69f1-11ea-92bb-e748f460b2f7.png)

SpaYco committed:
fixed rubocop

SpaYco committed:
added execution perm

SpaYco committed:
ignored execution error

SpaYco committed:
removed the ignore line

SpaYco committed:
fixed start_game not starting the game

SpaYco committed:
fixed game_start not running on linux terminal

SpaYco committed:
added file permissions

SpaYco commented:
turned out GitHub wasn't receiving file permissions from commits
also editing from windows gives some errors

euqueme commented on spec/game_spec.rb:
```diff
@@ -0,0 +1,20 @@
+require './bin/main.rb'
+
+RSpec.describe Game do
+  describe '#cehcker' do
```
- [ ] on line #6 I think it should say 'checker' instead of 'cehcker' let's correct it

euqueme commented on spec/game_spec.rb:
```diff
@@ -0,0 +1,20 @@
+RSpec.describe Game do
+  describe '#cehcker' do
+    it "returns game still on if the game isn't finished." do
+      checker_game_on = Game.new('SpaYco', 'Javier')
```
- [ ] Keep your code DRY. you used this variable 3 times in each test case declare it using [let](https://github.com/rubocop-hq/rspec-style-guide#let-blocks)

  > checker_game_on = Game.new('SpaYco', 'Javier')

euqueme commented on bin/main.rb:
```diff
@@ -0,0 +1,210 @@
+
+  def restart
+    start
+  end
```
- [ ] Let's create a test case for the restart method :muscle:

euqueme commented on bin/main.rb:
```diff
@@ -0,0 +1,210 @@
+    print 'would you like to play another game? (y/n) : '
+    answer = gets.chomp
+    answer == 'y' ? restart : (puts 'see you next time, goodbye!')
+  end
```
- [ ] Let's create a test case for the **start** method, to big to test? :thinking:  let's break it down in smaller methods so it's easier to test :muscle:

euqueme reviewed (changes requested):
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @javitocor and @Spayco  :balloon: :lollipop: 

### Your test cases work :medal_sports: we only need to add some more and consider negative test cases also remove some files not related to the project and address the comments I've left in your code bellow :point_down: 

## General

- [ ] Files not related to the project should not be committed and added to .gitignore ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

- remove the main file with no extension from your bin folder, you don't use it anywhere.

## Project Specific

- [ ] All the public methods of the classes implemented in project OOP are tested ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/all_tic_tac_toe.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- [ ] Specs for each method should be wrapped in one `describe` block. As in example here https://github.com/rubocop-hq/rspec-style-guide#empty-lines-around-examples . ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/describe.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- You have 5 public methods within your Game class so you should have 5 Describe blocks in your test cases 

- [ ] specs have to cover positive and negative scenarios ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/commonsense.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- please add one negative test case (using not_to) for each Describe block at least here is an example: 

```
RSpec.describe Game do
  describe checker_game_won do
    let(:checker) { Game.new('SpaYco', 'Javier')}
    it 'returns true if someone won' do
      expect(checker).to eql(false)
    end
    it 'doesn't return false if someone won' do
      expect(checker).not_to eql(false)
    end
  end
.
.
```

end

[OPTIONAL]

### Milestone 3 OOP requirements

#### Your project is playable but the code doesn't follow OOP practices and that's why is hard to build test cases for it so I strongly recommend the following changes to make it OOP cohesive (I'm sorry this wasn't pointed out to you when it should on Milestone 3) :point_down: 

- [ ] The project uses at least two classes (Usually it uses three: Game, Board and Player). The code of the project should be logically divided between classes. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/classes.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [ ] All classes are kept in `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- I know you have problems with your lib directory but even if you locate them in the bin directory make sure each class is in an independent file, you should only use the main.rb file to keep the game loop and call objects of the classes

- [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- Your start method does several things and it has the game loop, I strongly advise to move this loop out of the method and break the method down in smaller methods that can be easily tested

### Milestone 3 other requirements

#### These were also requested even though it won't help your project with OOP these are good programming practices so I also strongly recommend to have them

- [ ] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- The main.rb file shouldn't contain classes so all `puts/print` and `get.chomp`  shouldn't be within your Game class methods, your method classes should only retrieve objects (booleans, strings, integers, Games, Players, Boards etc.)

* [ ] Use attr_readers/attr_writers instead of attr_accessors. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- if you end up doing the Milestone 3 OOP changes it's good practice to either use attr_readers or attr_writers instead of attr_accessors  because sometimes you only need to be able to read an attribute instead of change it and it could stop cheaters to easily set values to your classes attributes

* [ ] DRY rule should be respected. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/dry.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- Don't Repeat Yourself, if you found yourself repeating code that looks similar within a method you could probably break it down so it's modular. for example, the code you have for printing the board could be done with a loop because it's repeating puts several times

this:
```
    puts '----------'
    puts "|#{$moves['a1']}|#{$moves['a2']}|#{$moves['a3']}|"
    puts '----------'
    puts "|#{$moves['b1']}|#{$moves['b2']}|#{$moves['b3']}|"
    puts '----------'
    puts "|#{$moves['c1']}|#{$moves['c2']}|#{$moves['c3']}|"
    puts '----------' 
```
could be achieved with a DRYer solution:

```
puts '----------'
['a', 'b', 'c'].each do |letter|
    3.times{ |number| print "|#{$moves["#{letter}#{number}"]}|"}
    puts "\n----------"
end
```
### That's it!, team :raised_hands: 

### Please :pray:  do the changes :man_technologist: :man_technologist: and submit for a code review when you are done :memo: 

Happy coding! :rocket: 

Any question? reach to me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

SpaYco committed:
Delete main

SpaYco committed:
added more tests

SpaYco committed:
Merge branch 'rspec' of github.com:Spayco/Tic-Tac-Toe-Game into rspec

javitocor committed:
Added spec for board class

javitocor committed:
FIxed rubocop errors

javitocor commented:
We have deleted the "main" file.
We have added a new class to control the board.
We split the main method into smaller ones.
We have added specs for all the public methods.

shshamim63 commented:
## Status :ambulance: Required Changes :ambulance: 
___
Hi @javitocor and @Spayco :wave:,
Wow a short pointed readme, contains most of the information that is needed to play the game. But the project structure does not seem like possessing all the requirements yet. Check out the points below and try to fix the issues.
___
We write tests to ensure that the individual units are performing properly. So as much as, you can test your classes and their methods, the more you are confirming that they are performing properly. But the way you have structured your project seems like, it is conflicting OOP principles. Your methods are tightly coupled. That is why is a struggling situation for you to add more test. It would be better for you if you schedule a meeting with a TSE to discuss more on this topic.
___
___
### Mandatory Changes
1. You can use let to assign a variable that will be used in your test. 
https://github.com/Spayco/Tic-Tac-Toe-Game/blob/8a6f3ec304cbfb2df546223004711e854cee96fe/spec/game_spec.rb#L6-L18
before line 7 you can add the following line:
```ruby
let(:checker_game_on) { Game.new('SpaYco', 'Javier') }
```
then you can remove the following line from the chunk.
checker_game_on = Game.new('SpaYco', 'Javier')
2. I agree with the previous TSE. Still, the number of methods you have tested are few.
___
### Optional Changes
___
1. All classes are kept in the lib directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
2. The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
___
##### Please fix the above-mentioned issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

javitocor committed:
Added specs for methods in String class

javitocor commented:
We have added specs for all the methods within String class
We have added the variable "let" to make the spec code better
All the public methods in classes string, board and game are tested
Every one of our methods does one single thing

whiz25 commented on spec/game_spec.rb:
```diff
@@ -0,0 +1,94 @@
+
+RSpec.describe Board do
+  describe '#show' do
+    board = Game.new('SpaYco', 'Javier')
```
- [x] use `let` here. `let(:board){Game.new...}`

whiz25 reviewed (changes requested):
## Nice Progress :+1: 
You have made good progress with the project since the last reviews. However, I can't approve this PR yet. We have a minor change to make:
- [x] make use of `let` to declare variables. Check my comment for suggestion.
 
### Martin Afani

javitocor commented:
Added "let" variable to board class specs

javitocor committed:
Added let variable to Board class

javitocor committed:
Fixed rubocop errors

whiz25 reviewed (approved):
## Congz :clap: 
All requested changes have been made. It's a :+1: from me.

### Martin Afani
